File: algorithm/mergesort.py
def merge(B, C, A):
    i = j = k = 0
    n = len(A)
    p = len(B)
    q = len(C)

    while i < p and j < q:
        if B[i] <= C[j]:
            A[k] = B[i]
            i += 1
        else:
            A[k] = C[j]
            j += 1
        k += 1
    
    if i == p:
        A[k : n] = C[j : q]
    else:
        A[k : n] = B[i : p]
        
def merge_sort(A):
    n = len(A)
    if n > 1:
        B = A[0 : n / 2]
        C = A[n / 2 : n]
        merge_sort(B)
        merge_sort(C)
        merge(B, C, A)

def main():
    A = [3, 6, 1, 6, 2, 6, 0, 9, 8]
    merge_sort(A)
    print A

if __name__ == '__main__':
    main()

File: algorithm/merge-sort.c
#include<stdio.h>
void merge(int *a, int *b, int *c, int na, int nb, int nc)
{
    int i = 0, j = 0, k = 0;
    while (i < na && j < nb) {
        if (a[i] <= b[j]) {
            c[k] = a[i];
            i++;
        } else {
            c[k] = b[j];
            j++;
        }
        k++;
    }
    if (i == na) {
        while (j < nb)
            c[k++] = b[j++];
    } else {
        while (i < na)
            c[k++] = a[i++];
    }
}

void merge_sort(int *num, int n)
{
    int i, na, nb, a[n], b[n];
    na = n / 2;
    nb = n - na;
    if (n > 1) {
        for (i = 0;i < na; i++) {
            a[i] = num[i];
        }
        for (i = na; i < n; i++) {
            b[i - na] = num[i];
        }

        merge_sort(a, na);
        merge_sort(b, nb);
        merge(a, b, num, na, nb, n);
    }
}

int main()
{
    int num[9] = {3, 6, 1, 6, 2, 6, 0, 9, 8};
    int i;
    merge_sort(num, 9);
    for (i = 0; i < 9; i++) {
        printf("%d ", num[i]);
    }
    printf("\n");

    return 0;
}


